src/routes.rs

```diff
@@ -1922,3 +2057,338 @@ pub(crate) async fn unlock(
+
+        // The cli takes the swaptype from the point of view of the user (=who wants to do the trade)
+        // In this function we're the market maker (=who sends the payment, completing the user trade)
+        // We swap the swaptype to hopefully make this function easier to read
+        let swaptype = swapstring.swap_type.opposite();
+
+        let receive_hints = unlocked_state.channel_manager
+            .list_usable_channels()
+            .iter()
+            .map(|details| {
+                let config = details.counterparty.forwarding_info.as_ref().unwrap();
+                RouteHint(vec![RouteHintHop {
+                    src_node_id: details.counterparty.node_id,
+                    short_channel_id: details.short_channel_id.unwrap(),
+                    cltv_expiry_delta: config.cltv_expiry_delta,
+                    htlc_maximum_msat: None,
+                    htlc_minimum_msat: None,
+                    fees: RoutingFees {
+                        base_msat: config.fee_base_msat,
+                        proportional_millionths: config.fee_proportional_millionths,
+                    },
+                }])
+            })
+            .collect();
+
+        let first_leg = get_route(
+            &unlocked_state.channel_manager,
+            &unlocked_state.router,
+            unlocked_state.channel_manager.get_our_node_id(),
+            taker_pk,
+            if swaptype.is_buy() { Some(swaptype.amount_msats()) } else { None },
+            if swaptype.is_buy() { None } else { Some(swapstring.asset_id) },
+            vec![],
+        );
+        let second_leg = get_route(
+            &unlocked_state.channel_manager,
+            &unlocked_state.router,
+            taker_pk,
+            unlocked_state.channel_manager.get_our_node_id(),
+            if swaptype.is_buy() { Some(HTLC_MIN_MSAT) } else { Some(swaptype.amount_msats()) },
+            if swaptype.is_buy() { Some(swapstring.asset_id) } else { None },
+            receive_hints,
+        );
+
+        let (mut first_leg, mut second_leg) = match (first_leg, second_leg) {
+            (Some(f), Some(s)) => (f, s),
+            (Some(_), _) => {
+                return Err(APIError::NoRoute);
+            }
+            (_, Some(_)) => {
+                return Err(APIError::NoRoute);
+            }
+            _ => {
+                return Err(APIError::NoRoute);
+            }
+        };
+
+        // Set swap flag
+        second_leg.paths[0].hops[0].short_channel_id |= IS_SWAP_SCID;
+
+        // Generally in the last hop the fee_amount is set to the payment amount, so we need to
+        // override it depending on what type of swap we are doing
+        if let SwapType::BuyAsset { .. } = swaptype {
+            first_leg.paths[0].hops.last_mut().expect("Path not to be empty").fee_msat = HTLC_MIN_MSAT;
+        } else {
+            first_leg.paths[0].hops.last_mut().expect("Path not to be empty").fee_msat = 0;
+        }
```

Comment:
When generating a route LDK treats the last `fee_msat` as a bit of a special value, because obviously a node receiving a payment isn't paying any more fees, there are no following hops, so it is normally set to the value of the whole payment.

In this case however the first `get_route` call only creates half of the route, and since we are extending it we need to override the last `fee_msat` of the first route. When the user is buying an asset it means they are receiving the asset via RGB, so the first half of the path that goes from the market maker to the user is an RGB payment. The rule is that RGB payments always pay `HTLC_MIN_MSAT` msats for fees, so here we override the last `fee_msat` with the correct value for an RGB payment

Reply:
thanks for explaining this. please note that this comment was also requesting a code change that is still pending